auth/rest.go

```diff
@@ -0,0 +1,180 @@
+// Copyright 2016 The Serviced Authors.
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package auth
+
+import (
+	"crypto/sha256"
+	"fmt"
+	"net/http"
+	"strings"
+	"time"
+
+	jwt "github.com/dgrijalva/jwt-go"
+)
+
+type AuthTokenRetriever func() (string, error)
+
+var (
+	RequestDelimiter                       = " "
+	RestTokenExpiration                    = 10 * time.Second
+	AuthTokenGetter     AuthTokenRetriever = AuthTokenNonBlocking
+)
+
+type RestToken interface {
+	Valid() error
+	Expired() bool
+	AuthToken() string
+	RestToken() string
+	RequestHash() []byte
+	HasAdminAccess() bool
+}
+
+type jwtRestClaims struct {
+	IssuedAt      int64  `json:"iat,omitempty"`
+	ExpiresAt     int64  `json:"exp,omitempty"`
+	DelegateToken string `json:"tkn,omitempty"`
+	ReqHash       []byte `json:"req,omitempty"`
+}
+
+func (t *jwtRestClaims) Valid() error {
+	if t.Expired() {
+		return ErrRestTokenExpired
+	}
+	return nil
+}
+
+func (t *jwtRestClaims) Expired() bool {
+	now := jwt.TimeFunc().UTC().Unix()
+	return now >= t.ExpiresAt
+}
+
+type jwtRestToken struct {
+	*jwtRestClaims
+	authIdentity Identity
+	restToken    string
+}
+
+func (t *jwtRestToken) AuthToken() string {
+	return t.DelegateToken
+}
+
+func (t *jwtRestToken) RequestHash() []byte {
+	return t.ReqHash
+}
+
+func (t *jwtRestToken) HasAdminAccess() bool {
+	return t.authIdentity.HasAdminAccess()
+}
+
+func (t *jwtRestToken) RestToken() string {
+	return t.restToken
+}
+
+func GetRequestHash(r *http.Request) []byte {
```

Comment:
Do we need this? Since the token is only valid for 10 seconds we should not worry about someone replaying it